man/rc-service.8
.\" Copyright (c) 2007-2015 The OpenRC Authors.
.\" See the Authors file at the top-level directory of this distribution and
.\" https://github.com/OpenRC/openrc/blob/master/AUTHORS
.\"
.\" This file is part of OpenRC. It is subject to the license terms in
.\" the LICENSE file found in the top-level directory of this
.\" distribution and at https://github.com/OpenRC/openrc/blob/master/LICENSE
.\" This file may not be copied, modified, propagated, or distributed
.\"    except according to the terms contained in the LICENSE file.
.\"
.Dd May 1, 2009
.Dt RC-SERVICE 8 SMM
.Os OpenRC
.Sh NAME
.Nm rc-service
.Nd locate and run an OpenRC service with the given arguments
.Sh SYNOPSIS
.Nm
.Op Fl i , -ifexists
.Ar service Op cmd...
.Nm
.Op Fl I , -ifinactive
.Ar service Op cmd...
.Nm
.Op Fl N , -ifnotstarted
.Ar service Op cmd...
.Nm
.Op Fl i , -ifexists
.Ar -a service
.Oo
.Ar cmd Op Ar arg...
.Oc
.Nm
.Op Fl I , -ifinactive
.Ar -a service
.Oo
.Ar cmd Op Ar arg...
.Oc
.Nm
.Op Fl N , -ifnotstarted
.Ar -a service
.Oo
.Ar cmd Op Ar arg...
.Oc
.Nm
.Fl e , -exists
.Ar service
.Nm
.Fl r , -resolve
.Ar service
.Nm
.Fl l , -list
.Sh DESCRIPTION
Service scripts could be in different places on different systems.
.Nm
locates the specified service and runs it with the given arguments.
If
.Fl i , -ifexists
is given then
.Nm
returns 0 even if the service does not exist.
If
.Fl I , -ifinactive
or
.Fl N , -ifnotstarted
is given then
.Nm
returns 0 if the service exists but is in the wrong state.
.Pp
When started (or stopped) from a runlevel, services can take only
a single service command. However, when run from the command line,
the service name can be followed by one or more service commands
(although there's typically only one). In addition, some service
commands (for some services) can take arguments, the presence of
which is indicated by the -a command line option. In this mode
only a single service command is supported.
.Pp
.Fl e , -exists
return 0 if it can find
.Ar service ,
otherwise -1.
.Fl r , -resolve
does the same and also prints the full path of the service to stdout.
.Pp
If given the
.Fl l , -list
argument then
.Nm
will list all available services.
.Sh SEE ALSO
.Xr openrc 8 ,
.Xr stdout 3
.Sh AUTHORS
.An Roy Marples <roy@marples.name>


src/rc/rc-service.c
/*
 * rc-service.c
 * Finds all OpenRC services
 */

/*
 * Copyright (c) 2008-2015 The OpenRC Authors.
 * See the Authors file at the top-level directory of this distribution and
 * https://github.com/OpenRC/openrc/blob/master/AUTHORS
 *
 * This file is part of OpenRC. It is subject to the license terms in
 * the LICENSE file found in the top-level directory of this
 * distribution and at https://github.com/OpenRC/openrc/blob/master/LICENSE
 * This file may not be copied, modified, propagated, or distributed
 *    except according to the terms contained in the LICENSE file.
 */

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "einfo.h"
#include "queue.h"
#include "rc.h"
#include "rc-misc.h"
#include "_usage.h"

const char *applet = NULL;
const char *extraopts = NULL;
const char *getoptstring = "+iINae:r:l" getoptstring_COMMON;
const struct option longopts[] = {
	{ "ifexists",     0, NULL, 'i' },
	{ "ifinactive",   0, NULL, 'I' },
	{ "ifnotstarted", 0, NULL, 'N' },
	{ "args",         0, NULL, 'a' },
	{ "exists",       1, NULL, 'e' },
	{ "resolve",      1, NULL, 'r' },
	{ "list",         0, NULL, 'l' },
	longopts_COMMON
};
const char * const longopts_help[] = {
	"if the service exists then run the command",
	"if the service is inactive then run the command",
	"if the service is not started then run the command",
	"service command is followed by args",
	"tests if the service exists or not",
	"resolve the service name to an init script",
	"list all available services",
	longopts_help_COMMON
};
const char *usagestring = ""
	"Usage: rc-service [options] [-i|-I|-N] <service> [<cmd>...]\n"
	"   or: rc-service [options] [-i|-I|-N] -a <service> [<cmd> [<arg>...]]\n"
	"   or: rc-service [options] -e <service>\n"
	"   or: rc-service [options] -r <service>\n"
	"   or: rc-service [options] -l\n";

int main(int argc, char **argv)
{
	int opt;
	char *service;
	RC_STRINGLIST *list;
	RC_STRING *s;
	RC_SERVICE state;
	bool if_exists = false;
	bool if_inactive = false;
	bool if_notstarted = false;
	bool has_args = false;

	applet = basename_c(argv[0]);
	/* Ensure that we are only quiet when explicitly told to be */
	unsetenv("EINFO_QUIET");

	while ((opt = getopt_long(argc, argv, getoptstring,
		    longopts, (int *) 0)) != -1)
	{
		switch (opt) {
		case 'e':
			service = rc_service_resolve(optarg);
			opt = service ? EXIT_SUCCESS : EXIT_FAILURE;
			free(service);
			return opt;
			/* NOTREACHED */
		case 'i':
			if_exists = true;
			break;
		case 'I':
			if_inactive = true;
			break;
		case 'N':
			if_notstarted = true;
			break;
		case 'l':
			list = rc_services_in_runlevel(NULL);
			if (TAILQ_FIRST(list) == NULL)
				return EXIT_FAILURE;
			rc_stringlist_sort(&list);
			TAILQ_FOREACH(s, list, entries)
			    printf("%s\n", s->value);
			rc_stringlist_free(list);
			return EXIT_SUCCESS;
			/* NOTREACHED */
		case 'r':
			service = rc_service_resolve(optarg);
			if (service == NULL)
				return EXIT_FAILURE;
			printf("%s\n", service);
			free(service);
			return EXIT_SUCCESS;
			/* NOTREACHED */
		case 'a':
			has_args = true;
			break;

		case_RC_COMMON_GETOPT
		}
	}

	argc -= optind;
	argv += optind;
	if (*argv == NULL)
		eerrorx("%s: you need to specify a service", applet);
	if ((service = rc_service_resolve(*argv)) == NULL) {
		if (if_exists)
			return 0;
		eerrorx("%s: service `%s' does not exist", applet, *argv);
	}
	state = rc_service_state(*argv);
	if (if_inactive && ! (state & RC_SERVICE_INACTIVE))
		return 0;
	if (if_notstarted && (state & RC_SERVICE_STARTED))
		return 0;
	if (has_args) {
		*argv-- = "-a";
		argc++;
	}
	*argv = service;
	execv(*argv, argv);
	eerrorx("%s: %s", applet, strerror(errno));
	/* NOTREACHED */
}
